css variables talk (#5)
deephaven-examples/lunch-n-learn: css-variables/readme.md

# CSS Variables for Deephaven Components

Our approach to css has been to have ~one and one match between components and css files.

This usually looks something like:

```
- MyComponent.js
- MyComponent.scss
```

Where CSS should always be nested/scoped to the component:

```scss
import './custom.scss'; // import global scss variables

$local-variable: $global-variable;

.my-component {
  .child {
    ...
    color: $local-variable;
  }
}
```

This is a good approach because it makes it easy to find the css for a component, and it makes it easy to see what css is being used by a component. And by scoping the css to the component, we avoid conflicts between components. We will continue to use this approach.

Historically, for variables we've used scss. With the adoption of spectrum we are switching to using css variables for our variables. CSS variables are great because they can be swapped out at runtime, which is necessary for themeing. They can be used almost anywhere you would have used a variable in scss.

> As we adopt spectrum, I expect the amount of custom css one has to write to continually go down. For example, the whole ACL editor has very little custom css. (14 styles, and I bet we could actually reduce that by half or more, as some could be props on spectrum components).

Reviewer: ```suggestion
This is a good approach because it makes it easy to find the CSS for a component, and it makes it easy to see what CSS is being used by a component. And by scoping the CSS to the component, we avoid conflicts between components. We will continue to use this approach.

Historically, for variables we've used SCSS. With the adoption of spectrum we are switching to using CSS variables for our variables. CSS variables are great because they can be swapped out at runtime, which is necessary for theming. They can be used almost anywhere you would have used a variable in SCSS.

As we adopt spectrum, I expect the amount of custom CSS one has to write to continually go down. For example, the whole ACL editor has very little custom CSS. (14 styles, and I bet we could actually reduce that by half or more, as some could be props on spectrum components).
```

## Defining CSS variables

They are defined with a `--` prefix, and used with `var(--variable-name)`. See: [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties)

```scss
$color: red;
// vs

> <some scope> { // :root is the global scope, or a css selector

Reviewer: ```suggestion
<some scope> { // :root is the global scope, or a CSS selector
```

--color: red;
}

.my-component {
  color: $color;
  color: var(--color);
}
```

```scss
:root {
  --a: 1;
  --b: 2;
  // can also contain numbers, and be used in calc
  --result: calc(var(--a) + var(--b) * 1px); // apply a unit
  // --result: 3;

  --color: invalid;
  // variable is ignored if used as a color property
  

.my-component {
  // can also specify fallback values if a value is not defined
  color: var(--my-undefined-var, red);
}
```

Fallback values to expose variables:

```scss

> // spectrum uses this pattern for themeing, where the component defines a variable and a fallback value

Reviewer: ```suggestion
// spectrum uses this pattern for theming, where the component defines a variable and a fallback value
```

// and the user can override it by setting the undefined variable in a theme
.my-spectrum-component {
  // exposes --my-spectrum-component-color as a themeable variable, but defaults to a fallback value
  // kinda pollutes the global scope
  color: var(--my-spectrum-component-color, var(--global-color));
}

// then in theme
:root {
  --my-spectrum-component-color: blue;
}
```

vs. 

```scss
.my-spectrum-component {
  // encapsulates the variable in the component scope
  --my-spectrum-component-color: var(--global-color);
  color: var(--my-spectrum-component-color);
}

// then in theme
:root {
  --my-spectrum-component-color: blue; // wouldn't work because of specificity
}
.my-spectrum-component {
  --my-spectrum-component-color: blue; // does work
}

```

> Consider this example of a button component with css variables and variants:

Reviewer: ```suggestion
Consider this example of a button component with CSS variables and variants:
```

}

.btn-negative {
  --btn-bg-color: var(--negative-color);
  --btn-fg-color: var(--negative-text-color);
}
```

```jsx
<button class="btn btn-primary"/>
<button class="btn btn-negative"/>
```

## Inheritance

CSS variables can be defined either in the root scope, or in a selector. We **use the root scope for global theme variables**, and in **the selector for component specific variables**. Inheritance works as expected, so if a component doesn't define a variable, it will inherit the value from the root scope. Use fallbacks pattern, in the case of something you are considering a semi-private API, and prefix with `--dh` to avoid conflicts (global).

```scss
// global theme
:root {
  --global-variable: red;
}
```

```scss
// MyComponent.scss
.my-component {
  --local-color: var(--global-variable);

  .child {
    color: var(--local-color);
  }

  .child.a {
    --local-color: salmon;
  }

  .child.b {
    // dh-child-b could inherit from the root scope, and thought of as a semi-private API
    // in the sense that it's not published in the theme files, but it is in the global scope
    --local-color: var(--dh-child-b, lemonchiffon);
  }
}
```

## Variables in JS

> A challenge with that is, monaco, ploly and our grid all require resolved javascript values for theming. We accomplish this by using JS to read the value of each css variable defined in our theme (for numbers, it getProperty, and colors by setting a background and reading the color back from the element).

Reviewer: ```suggestion
A challenge with using inheritance is that monaco, plotly and our grid all require resolved javascript values for theming. We accomplish this by using JS to read the value of each CSS variable defined in our theme (for numbers, it getProperty, and colors by setting a background and reading the color back from the element).
```

```js
:root {
  --color: red;
}

const propetry = getComputedStyle(document.documentElement).getPropertyValue('--color');
// "red"
const div = document.createElement('div');
div.style.backgroundColor = 'var(--color)';
const resolved_color = getComputedStyle(div).style.backgroundColor;
// "rgb(255, 0, 0)" which we then normalize with JS to hex for use in monaco, plotly, and our grid.
```

Aside CSS properties are actually a great way to get values _from_ JS into CSS:

[Codepen Example](https://codepen.io/dsmmcken/pen/ZEPjBmy)

```js
document.addEventListener("mousemove", (e) => {
  document.documentElement.style.setProperty("--x", e.clientX);
  document.documentElement.style.setProperty("--y", e.clientY);
});
```

```scss
div {
  transform: translate(var(--x) * 1px, var(--y) * 1px);
}
```

> ## Start using css variables!

Reviewer: ```suggestion
## Start using CSS variables!
```

> We have a theme component which manages which theme is applied. All users created themes must inheriet from either a light or a dark theme.

Reviewer: ```suggestion
We have a theme component which manages which theme is applied. All users created themes must inherit from either a light or a dark theme.
```

> 1. All Deephaven **theme variables** start with `--dh` to avoid conflicts with other css variables. `--dh` variables should be thought of as a public API. Removing a `--dh` variable is a breaking change. Adding a variable is not a breaking change, as users themes inherit from the base themes, and will not break if a variable is added.

Reviewer: ```suggestion
1. All Deephaven **theme variables** start with `--dh` to avoid conflicts with other CSS variables. `--dh` variables should be thought of as a public API. Removing a `--dh` variable is a breaking change. Adding a variable is not a breaking change, as users themes inherit from the base themes, and will not break if a variable is added.
```

 
   Local variables to components should not start with `--dh`, but instead should follow the component name `--my-component`.

> 2. A theme consists of a set of css variables defined across a number of files. Variables added to Dark or Light theme **must be added to both base theme files**. You will rarely need to add a variable to a theme file unless we think it should be part of the public api.

Reviewer: ```suggestion
2. A theme consists of a set of CSS variables defined across a number of files. Variables added to Dark or Light theme **must be added to both base theme files**. You will rarely need to add a variable to a theme file unless we think it should be part of the public API.
```

3. Components should use the most specific available variable. If a **semantic _component_ variable** is available use that, otherwise it should use a **semantic variable** and lastly it should use a **palette variable**.

4. Components should almost never use palette variables directly.

   <small>Sometimes maybe a gray-*, but really mostly no.</small>

5. Components can and should define their own internal variables, scoped to the component.

Please review the /styleguide for a complete list of exposed variables.

### Stop using SCSS color variables!

> Immediately discontinue use of scss color variables in favor of css variables.

Reviewer: ```suggestion
Immediately discontinue use of SCSS color variables in favor of CSS variables.
```

- scss unit variables like $spacer-*, or $border-radius are still ok to use (for now).

> - All bootstrap variables have been re-defined to use the css variables, but we haven't removed the scss variables yet. They should be considered deprecated.

Reviewer: ```suggestion
- All bootstrap variables have been re-defined to use the CSS variables, but we haven't removed the scss variables yet. They should be considered deprecated.
```

  ex. 
  ```scss
  $content-bg: var(--dh-color-content-bg);
  $primary: var(--dh-color-accent);
  ```

> - Immediately **stop using compile time color manipulation functions in scss**. This is no longer possible darken($primary, 10%) does not work against a runtime variable. However, good news is with the expanded palette, this is no longer necessary. For ex. Accent 100-1400 covers a much wider range of colors than we had before. All existing scss color manipulations were removed and replaced as part of themeing work.

Reviewer: ```suggestion
- Immediately **stop using compile time color manipulation functions in scss**. This is no longer possible darken($primary, 10%) does not work against a runtime variable. However, good news is with the expanded palette, this is no longer necessary. For ex. Accent 100-1400 covers a much wider range of colors than we had before. All existing SCSS color manipulations were removed and replaced as part of theming work.
```

> - Transparency is a special case. Since user supplied variables can be in any sRGB formart we must also stop using scss functions for transparency. Instead of `rgba($color, 0.5)` use either a color-mix function:

Reviewer: ```suggestion
- Transparency is a special case. Since user supplied variables can be in any sRGB format we must also stop using SCSS functions for transparency. Instead of `rgba($color, 0.5)` use either a color-mix function:
```

    // --dh-color-visual-negative: var(--dh-color-red-800);
    // --dh-color-visual-notice: var(--dh-color-yellow-1200);
    // --dh-color-visual-info: var(--dh-color-cyan-1100);

 // applies palette and semantic variables to global component variables
 /theme-dark-components.scss 
    // Limited set of specific component variables to be used contextually.
    // Like if you are making a new custom input component, consider:
    // I expect this file to grow slightly over time

    // --dh-color-input-bg: var(--dh-color-gray-75);
    // --dh-color-input-fg: var(--dh-color-text);
    // --dh-color-input-border: var(--dh-color-border);

 // applies variables to indepentently themed components
 /theme-dark-semantic-grid.scss
    // used exclusively by iris-grid, not for use in other components
    // --dh-color-grid-* variables
 /theme-dark-semantic-editor.scss
     // used exclusively by monaco, not for use in other components
    // --dh-color-editor-* variables
 /theme-dark-semantic-chart.scss
    // used exclusively by plotly, not for use in other components
    // --dh-color-chart-* variables
```

## Palette Variables

For reference only, do not use directly. 

> Light and dark invert there range of colors from 50-900. ie. 900 is black in light theme and white in dark theme.

Reviewer: ```suggestion
Light and dark invert their range of colors from 50-900. ie. 900 is black in light theme and white in dark theme.
```

Things like input states, or action button states, or item-list hover/focus states. Use these if you are creating custom compoents, like a new type of input or picker component.

Variables can and will be added here for things that we think should be promoted to be part of the Public API. For example, golden layout has a handful of additional variables necessary for theme development, defined as semantic component variables.

![](./semantic-components.jpg)

```scss
--dh-color-input-* // used for custom input components
--dh-color-action-*  // used for custom button or action components
--dh-color-selector-* // used for custom picker/selector components
--dh-color-item-list-* // used for custom item list components
```

## Example

For example, let's say you made a new color picker selector:

```scss
// A square with the selected color as a fill, and a border
// use the semantic input variables for the border
.my-color-selector {
    --active-color: var(--default-color, var(--dh-color-accent));

    height: 16px;
    width: 16px;
    background-color: var(--active-color);
    border: 1px solid var(--dh-color-input-border); // re-use the semantic variable

>     border-radius: $border-radius; // non-colors variables are scss variables for now

Reviewer: ```suggestion
    border-radius: $border-radius; // non-colors variables are SCSS variables for now
```
